src/object_store/src/object/s3.rs

```diff
@@ -242,22 +224,9 @@ impl StreamingUploader for S3StreamingUploader {
         fail_point!("s3_finish_streaming_upload_err", |_| Err(
             ObjectError::internal("s3 finish streaming upload error")
         ));
-        // Fallback to `PUT`.
-        if self.join_handles.is_empty() {
+        if self.join_handles.is_empty() && self.buf.is_empty() {
             self.abort().await?;
```

Comment:
Have we considered lazily starting multi-part upload when we have the first part to upload so that we can avoid unnecessary start and abort?

Reply:
Abort will no longer be necessary because even if the object size is < 5MiB, we can still `CompleteMultipartUpload`, because there's no minimum size limit on the last part.

Reply:
If the object size is < part size, there will only be one part in the upload. If we do multi-part upload, there will be 3 requests, start, upload and complete. If we only do a put, there will only be a put request. Lazily starting multi-part can enable this optimization.

But it seems irrelevant to this PR. We may support this in future PR.